pd-client/client.go

```diff
@@ -43,77 +50,279 @@ type Client interface {
 	Close()
 }
 
+type tsoRequest struct {
+	done     chan error
+	physical int64
+	logical  int64
+}
+
+const (
+	pdTimeout             = 3 * time.Second
+	maxMergeTSORequests   = 10000
+	maxInitClusterRetries = 100
+)
+
 type client struct {
-	worker *rpcWorker
+	urls        []string
+	clusterID   uint64
+	tsoRequests chan *tsoRequest
+
+	mu            sync.RWMutex
+	clients       map[string]pdpb2.PDClient
+	leader        string
+	checkLeaderCh chan struct{}
+
+	wg   sync.WaitGroup
+	quit chan struct{}
 }
 
 // NewClient creates a PD client.
 func NewClient(pdAddrs []string) (Client, error) {
 	log.Infof("[pd] create pd client with endpoints %v", pdAddrs)
-	worker, err := newRPCWorker(pdAddrs)
-	if err != nil {
+	c := &client{
+		urls:          addrsToUrls(pdAddrs),
+		tsoRequests:   make(chan *tsoRequest, maxMergeTSORequests),
+		clients:       make(map[string]pdpb2.PDClient),
+		checkLeaderCh: make(chan struct{}, 1),
+		quit:          make(chan struct{}),
+	}
+
+	if err := c.initClusterID(); err != nil {
 		return nil, errors.Trace(err)
 	}
-	return &client{worker: worker}, nil
+	if err := c.updateLeader(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	log.Infof("[pd] init cluster id %v", c.clusterID)
+
+	c.wg.Add(2)
+	go c.tsLoop()
+	go c.leaderLoop()
+
+	// TODO: Update addrs from server continuously by using GetMember.
+
+	return c, nil
+}
+
+func (c *client) initClusterID() error {
+	for i := 0; i < maxInitClusterRetries; i++ {
+		for _, u := range c.urls {
+			client, err := apiutil.NewClient(u, pdTimeout)
```

Comment:
I see, because `GetPDMembers` is not implemented yet. How about adding a TODO?

BTW, I have implemented a full PD service, see more https://github.com/pingcap/pd/blob/28fb71e8d8a010162bab85524623342c0251e964/server/api/v2rpc/server.go